views-react/src/main/java/io/micronaut/views/react/ReactJSBeanFactory.java

```diff
@@ -0,0 +1,138 @@
+import io.micronaut.views.react.util.JavaUtilLoggingToSLF4J;
+import io.micronaut.views.react.util.OutputStreamToSLF4J;
+import jakarta.inject.Named;
+import jakarta.inject.Singleton;
+import org.graalvm.polyglot.Engine;
+import org.graalvm.polyglot.HostAccess;
+import org.graalvm.polyglot.SandboxPolicy;
+import org.graalvm.polyglot.Source;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.event.Level;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Optional;
+
+import static java.lang.String.format;
+
+/**
+ * Allows the default Javascript context and host access policy to be controlled.
+ */
+@Factory
+@Internal
+final class ReactJSBeanFactory {
+    static final String REACT_QUALIFIER = "react";
+
+    private static final Logger LOG = LoggerFactory.getLogger("js");
+
+    // We cache the Source objects because they are expensive to create, but, we don't want them
+    // to be singleton beans so we can recreate them on file change.
+    private Source serverBundle;  // L(this)
+    private Source renderScript;  // L(this)
+
+    /**
+     * This defaults to {@link HostAccess#ALL} if the sandbox is disabled, or {@link
+     * HostAccess#CONSTRAINED} if it's on. By replacing the {@link HostAccess} bean you can
+     * whitelist methods/properties by name or annotation, which can be useful for exposing third
+     * party libraries where you can't add the normal {@link HostAccess.Export} annotation, or
+     * allowing sandboxed JS to extend or implement Java types.
+     */
+    @Singleton
+    @Named(REACT_QUALIFIER)
+    HostAccess hostAccess(ReactViewsRendererConfiguration configuration) {
+        return configuration.getSandbox()
+            ? HostAccess.newBuilder(HostAccess.CONSTRAINED).allowListAccess(true).allowMapAccess(true).build()
+            : HostAccess.ALL;
+    }
+
+    @Singleton
+    BeanPool<ReactJSContext> contextPool(ApplicationContext applicationContext) {
+        return new BeanPool<>(() -> applicationContext.createBean(ReactJSContext.class));
+    }
+
+    @Singleton
+    @Named(REACT_QUALIFIER)
+    Engine engine(ReactViewsRendererConfiguration configuration) {
+        boolean sandbox = configuration.getSandbox();
+        LOG.debug("ReactJS sandboxing {}", sandbox ? "enabled" : "disabled");
+        return Engine.newBuilder("js")
+            .out(new OutputStreamToSLF4J(LOG, Level.INFO))
+            .err(new OutputStreamToSLF4J(LOG, Level.ERROR))
+            .logHandler(new JavaUtilLoggingToSLF4J(LOG))
+            .sandbox(sandbox ? SandboxPolicy.CONSTRAINED : SandboxPolicy.TRUSTED)
+            .build();
+    }
+
+    @Bean
+    @Named(REACT_QUALIFIER)
+    synchronized Source serverBundle(ResourceResolver resolver, ReactViewsRendererConfiguration reactConfiguration) throws IOException {
```

Comment:
Why do you need `synchronized`? If the bean is a singleton it should be created only once

Reply:
They aren't singletons, they can be invalidated by file changes on disk and re-created in response to that event (see the code in `maybeReloadServerBundle`). When that happens the bean pool is flushed and the contexts are recreated.